src/test/java/com/gregtechceu/gtceu/api/recipe/ingredient/IntProviderFluidIngredientTest.java

```diff
@@ -0,0 +1,975 @@
+        int runs = 7;
+        fluidIn.setFluidInTank(0, new FluidStack(CR_OUT, runs));
+        // 1t to turn on, 2t per recipe run
+        // get the result of each roll independently
+        int[] addedRolls = new int[runs];
+
+        helper.runAfterDelay(4, () -> {
+            if (machine.getRecipeLogic().getLastRecipe().getOutputContents(FluidRecipeCapability.CAP).get(0)
+                    .getContent() instanceof IntProviderFluidIngredient ingredient) {
+                ingredient.setSampledCount(0);
+
+                if (ingredient.getSampledCount() != 0) {
+                    helper.fail("Singleblock Ranged Fluid Output sabotage failed! " +
+                            "Output count not was altered!");
+                }
+            } else {
+                helper.fail("Singleblock Ranged Fluid Output sabotage failed! " +
+                        "Recipe logic did not contain a Ranged Output!");
+            }
+        });
+        for (int i = 0; i < runs; i++) {
+            final int finalI = i; // lambda preserve you
+            helper.runAfterDelay(2 * i + 3, () -> {
+                addedRolls[finalI] = (int) fluidOut.getTotalContentAmount();
+            });
+        }
+        // check the results of all rolls together
+        helper.runAfterDelay(runs * 2 + 1, () -> {
+            FluidStack results = fluidOut.getFluidInTank(0);
+            helper.assertTrue(TestUtils.isFluidWithinRange(results, runs, runs * 9),
+                    "Sabotaged Singleblock CR didn't produce correct number of fluids, produced [" +
+                            results.getAmount() + "] not [" + runs + "-" + (runs * 9) + "]");
+            helper.assertFalse((results.getAmount() == runs * 9),
+                    "Sabotaged Singleblock CR rolled max value on every roll (how??)");
+            helper.assertFalse((results.getAmount() == runs * 0),
+                    "Sabotaged Singleblock CR rolled min value on every roll! " +
+                            "This is the failure this sabotage was intended to induce.");
+
+            // check if all the rolls were equal, but not min/max
+            int[] rolls = new int[runs];
+            rolls[0] = addedRolls[0];
+            boolean allEqual = false;
+            for (int i = 1; i < runs; i++) {
+                rolls[i] = addedRolls[i] - addedRolls[i - 1];
+                if (rolls[i] == rolls[i - 1]) {
+                    allEqual = true;
+                } else {
+                    allEqual = false;
+                    break;
+                }
+            }
+            helper.assertFalse(allEqual,
+                    "Sabotaged Singleblock CR rolled the same value on every input roll (rolled " + rolls[0] + ")");
+            helper.succeed();
+        });
+    }
+
+    // Failure Test for singleblock machine with ranged fluid input
+    // Provides too little input fluid, should not run recipes.
+    @GameTest(template = "singleblock_charged_cr", batch = "RangedIngredients")
```

Comment:
Some of these are wrong, make sure it's fluid for fluid and nothing for items :)